ext/random/random.c

```diff
@@ -0,0 +1,1762 @@
+}
+/* }}} */
+
+/* {{{ Unserialize object */
+PHP_METHOD(Random_Engine_CombinedLCG, __unserialize)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	HashTable *data;
+	zval *tmp;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_HT(data);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* members */
+	tmp = zend_hash_index_find(data, 0);
+	if (!tmp || Z_TYPE_P(tmp) != IS_ARRAY) {
+		zend_throw_exception(NULL, "Incomplete or ill-formed serialization data", 0);
+		RETURN_THROWS();
+	}
+	object_properties_load(&engine->std, Z_ARRVAL_P(tmp));
+
+	/* state */
+	tmp = zend_hash_index_find(data, 1);
+	if (!tmp || Z_TYPE_P(tmp) != IS_ARRAY) {
+		zend_throw_exception(NULL, "Incomplete or ill-formed serialization data", 0);
+		RETURN_THROWS();
+	}
+	if (engine->algo->unserialize(engine->state, Z_ARRVAL_P(tmp)) == FAILURE) {
+		zend_throw_exception(NULL, "Engine unserialize failed", 0);
+		RETURN_THROWS();
+	}
+}
+/* }}} */
+
+/* {{{ DebugInfo */
+PHP_METHOD(Random_Engine_CombinedLCG, __debugInfo)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	zval tmp;
+
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	if (!engine->std.properties) {
+		rebuild_object_properties(&engine->std);
+	}
+	ZVAL_ARR(return_value, zend_array_dup(engine->std.properties));
+
+	if (engine->algo->serialize) {
+		array_init(&tmp);
+		if (engine->algo->serialize(engine->state, Z_ARRVAL(tmp)) == FAILURE) {
+			zend_throw_exception(NULL, "Engine serialize failed", 0);
+			RETURN_THROWS();
+		}
+		zend_hash_str_add(Z_ARR_P(return_value), "__states", sizeof("__states") - 1, &tmp);
+	}
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Engine_MersenneTwister, __construct)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	php_random_engine_state_mersennetwister *state = engine->state;
+	zend_long seed, mode = MT_RAND_MT19937;
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_LONG(seed);
+		Z_PARAM_OPTIONAL;
+		Z_PARAM_LONG(mode)
+	ZEND_PARSE_PARAMETERS_END();
+
+	switch (mode) {
+		case MT_RAND_PHP:
+			state->mode = MT_RAND_PHP;
+			break;
+		default:
+			state->mode = MT_RAND_MT19937;
+	}
+
+	engine->algo->seed(engine->state, seed);
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Engine_Secure, __construct)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Engine_XorShift128Plus, __construct)
+{
+	php_random_engine *engine = Z_RANDOM_ENGINE_P(ZEND_THIS);
+	php_random_engine_state_xorshift128plus *state = engine->state;
+	zend_string *str_seed = NULL;
+	zend_long int_seed = 0;
+	int i, j;
+	
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_STR_OR_LONG(str_seed, int_seed)
+	ZEND_PARSE_PARAMETERS_END();
+	
+	if (str_seed) {
+		/* char (8 bit) * 16 = 128 bits */
+		if (str_seed->len == 16) {
+			/* Endianness safe copy */
+			for (i = 0; i < 2; i++) {
+				for (j = 0; j < 8; j++) {
+					state->s[i] += (unsigned char) ZSTR_VAL(str_seed)[(i * j) + j] >> (j * 8);
+				}
+			}
```

Comment:
```suggestion
			for (i = 0; i < 2; i++) {
				state->s[i] = 0;
				for (j = 0; j < 8; j++) {
					state->s[i] += ((uint64_t) (unsigned char) ZSTR_VAL(str_seed)[(i * 8) + j]) << (j * 8);
				}
			}
```

a) Missing reset of the state fields to 0.
b) Missing cast to uint64_t.
c) (i * j) must be (i * 8).

Same for Random_Engine_Xoshiro256StarStar further below.

Reply:
You should probably add a simple test for a string seed. You can take this:

```
<?php

use Random\Engine\XorShift128Plus;

$g = new XorShift128Plus("\x01\x02\x03\x04\x05\x06\x07\x08\x01\x02\x03\x04\x05\x06\x07\x08");

var_dump($g);
/*
object(Random\Engine\XorShift128Plus)#1 (1) {
  ["__states"]=>
  array(2) {
    [0]=>
    string(18) "578437695752307201"
    [1]=>
    string(18) "578437695752307201"
  }
}
*/

for ($i = 0; $i < 1000; $i++) $g->generate();
\var_dump(\bin2hex($g->generate())); // string(16) "c05386f47f7b6b51"
```
`578437695752307201` is `0x0807060504030201` in hexadecimal, thus this matches the little endian parsing.

Reply:
add test: d31ab902100c0a6d7e024aa0c4c34c5d0802b780